File: src/ocaml/flux_constraint.ml
// FLUX Constraint Engine — OCaml 5
// Pure INT8 saturated constraint checking. Zero dependencies.

let int8_min = -127
let int8_max = 127

let saturate v = max int8_min (min int8_max v)

type severity = Pass | Caution | Warning | Critical

type constraint_def = { lo: int; hi: int; name: string }

type flux_result = {
  error_mask: int;
  severity: severity;
  violated_lo: int;
  violated_hi: int;
  violated_count: int;
  passed: bool;
}

let check constraints value =
  let val = saturate value in
  let error_mask, violated_lo, violated_hi, violated_count =
    List.fold_left (fun (em, vlo, vhi, vc) (i, c) ->
      let lo = saturate c.lo and hi = saturate c.hi in
      let lo_fail = val < lo and hi_fail = val > hi in
      let em' = if lo_fail || hi_fail then em lor (1 lsl i) else em in
      let vlo' = if lo_fail then vlo lor (1 lsl i) else vlo in
      let vhi' = if hi_fail then vhi lor (1 lsl i) else vhi in
      let vc' = if lo_fail || hi_fail then vc + 1 else vc in
      (em', vlo', vhi', vc')
    ) (0, 0, 0, 0) (List.mapi (fun i c -> (i, c)) constraints)
  in
  let nc = List.length constraints in
  let sev = if violated_count = 0 then Pass
            else if violated_count <= nc / 4 then Caution
            else if violated_count <= nc / 2 then Warning
            else Critical in
  { error_mask; severity = sev; violated_lo; violated_hi; violated_count; passed = sev = Pass }

let check_batch constraints values =
  let results = List.map (check constraints) values in
  let count_sev s = List.length (List.filter (fun r -> r.severity = s) results) in
  let stats = [
    ("pass", count_sev Pass); ("caution", count_sev Caution);
    ("warning", count_sev Warning); ("critical", count_sev Critical)
  ] in
  (results, stats)

let benchmark constraints iterations =
  let t0 = Sys.time () in
  for _ = 1 to iterations do
    ignore (check constraints 50)
  done;
  let elapsed = Sys.time () -. t0 in
  (float iterations *. float (List.length constraints) /. elapsed, elapsed *. 1000.0)

let presets = [
  ("aviation", [
    {lo = -55; hi = 70; name = "cabin_temp_C"};
    {lo = 75; hi = 101; name = "cabin_pressure_kPa"};
    {lo = 0; hi = 100; name = "fuel_flow_pct"};
    {lo = 60; hi = 100; name = "hydraulic_pct"};
  ]);
  ("medical", [
    {lo = 36; hi = 38; name = "body_temp_C"};
    {lo = 60; hi = 100; name = "heart_rate_bpm"};
    {lo = 95; hi = 100; name = "spo2_pct"};
    {lo = 80; hi = 120; name = "bp_systolic_mmHg"};
  ]);
  ("maritime", [
    {lo = -2; hi = 35; name = "sea_temp_C"};
    {lo = 50; hi = 100; name = "hull_integrity_pct"};
    {lo = 0; hi = 50; name = "wave_height_m"};
    {lo = 0; hi = 80; name = "wind_speed_kn"};
  ]);
  ("automotive", [
    {lo = -40; hi = 60; name = "battery_temp_C"};
    {lo = 0; hi = 100; name = "soc_pct"};
    {lo = 0; hi = 100; name = "charge_rate_pct"};
    {lo = 20; hi = 80; name = "cabin_temp_C"};
  ]);
  ("energy", [
    {lo = 49; hi = 51; name = "grid_freq_Hz_x10"};
    {lo = 95; hi = 105; name = "voltage_pct"};
    {lo = 0; hi = 80; name = "transformer_temp_C"};
    {lo = 0; hi = 100; name = "line_load_pct"};
  ]);
]

let from_preset name =
  match List.assoc_opt name presets with
  | Some cs -> cs
  | None -> failwith (Printf.sprintf "Unknown preset: %s" name)

(* Self-test *)
let () =
  Printf.printf "FLUX Constraint Engine — OCaml\n";
  Printf.printf "===============================\n";
  assert (saturate (-128) = -127);
  assert (saturate 128 = 127);
  Printf.printf "  saturate: OK\n";
  let fc = [{lo = 0; hi = 100; name = "test"}] in
  assert (check fc 50).passed;
  assert (not (check fc 150).passed);
  Printf.printf "  check: OK\n";
  let fc4 = [{lo=0;hi=10;name="a"};{lo=0;hi=10;name="b"};{lo=0;hi=10;name="c"};{lo=0;hi=10;name="d"}] in
  let r = check fc4 50 in
  assert (r.severity = Critical && r.violated_count = 4);
  Printf.printf "  severity: OK\n";
  let fc3 = from_preset "aviation" in
  assert (List.length fc3 = 4);
  Printf.printf "  presets: OK\n";
  Printf.printf "  All tests pass\n"

File: src/fsharp/FluxConstraint.fs
// FLUX Constraint Engine — F#
// Pure INT8 saturated constraint checking. Zero dependencies.

module FluxConstraint

let Int8Min, Int8Max = -127, 127

let saturate v = max Int8Min (min Int8Max v)

type Severity = Pass = 0 | Caution = 1 | Warning = 2 | Critical = 3

type ConstraintDef = { Lo: int; Hi: int; Name: string }

type FluxResult = {
    ErrorMask: int
    Severity: Severity
    ViolatedLo: int
    ViolatedHi: int
    ViolatedCount: int
    Passed: bool
}

let check (constraints: ConstraintDef list) (value: int) =
    let val = saturate value
    let mutable errorMask, violatedLo, violatedHi, violatedCount = 0, 0, 0, 0

    constraints |> List.iteri (fun i c ->
        let lo = saturate c.Lo
        let hi = saturate c.Hi
        let loFail = val < lo
        let hiFail = val > hi
        if loFail || hiFail then
            errorMask <- errorMask ||| (1 <<< i)
            violatedCount <- violatedCount + 1
        if loFail then violatedLo <- violatedLo ||| (1 <<< i)
        if hiFail then violatedHi <- violatedHi ||| (1 <<< i)
    )

    let nc = constraints.Length
    let sev =
        if violatedCount = 0 then Severity.Pass
        elif violatedCount <= nc / 4 then Severity.Caution
        elif violatedCount <= nc / 2 then Severity.Warning
        else Severity.Critical

    { ErrorMask = errorMask; Severity = sev; ViolatedLo = violatedLo
      ViolatedHi = violatedHi; ViolatedCount = violatedCount; Passed = sev = Severity.Pass }

let checkBatch constraints values =
    let results = values |> List.map (check constraints)
    let stats = Map [
        "pass", results |> List.filter (fun r -> r.Severity = Severity.Pass) |> List.length
        "caution", results |> List.filter (fun r -> r.Severity = Severity.Caution) |> List.length
        "warning", results |> List.filter (fun r -> r.Severity = Severity.Warning) |> List.length
        "critical", results |> List.filter (fun r -> r.Severity = Severity.Critical) |> List.length
    ]
    results, stats

let benchmark constraints iterations =
    let sw = System.Diagnostics.Stopwatch.StartNew()
    for i = 0 to iterations - 1 do
        check constraints ((i % 254) - 127) |> ignore
    sw.Stop()
    let rate = float iterations * float constraints.Length / sw.Elapsed.TotalSeconds
    rate, sw.Elapsed.TotalMilliseconds

let presets = Map [
    "aviation", [
        { Lo = -55; Hi = 70; Name = "cabin_temp_C" }
        { Lo = 75; Hi = 101; Name = "cabin_pressure_kPa" }
        { Lo = 0; Hi = 100; Name = "fuel_flow_pct" }
        { Lo = 60; Hi = 100; Name = "hydraulic_pct" }
    ]
    "medical", [
        { Lo = 36; Hi = 38; Name = "body_temp_C" }
        { Lo = 60; Hi = 100; Name = "heart_rate_bpm" }
        { Lo = 95; Hi = 100; Name = "spo2_pct" }
        { Lo = 80; Hi = 120; Name = "bp_systolic_mmHg" }
    ]
    "automotive", [
        { Lo = -40; Hi = 60; Name = "battery_temp_C" }
        { Lo = 0; Hi = 100; Name = "soc_pct" }
        { Lo = 0; Hi = 100; Name = "charge_rate_pct" }
        { Lo = 20; Hi = 80; Name = "cabin_temp_C" }
    ]
    "energy", [
        { Lo = 49; Hi = 51; Name = "grid_freq_Hz_x10" }
        { Lo = 95; Hi = 105; Name = "voltage_pct" }
        { Lo = 0; Hi = 80; Name = "transformer_temp_C" }
        { Lo = 0; Hi = 100; Name = "line_load_pct" }
    ]
]

let fromPreset name =
    match presets.TryFind name with
    | Some cs -> cs
    | None -> failwith $"Unknown preset: {name}"

[<EntryPoint>]
let main _ =
    printfn "FLUX Constraint Engine — F#"
    printfn "==========================="

    assert (saturate -128 = -127)
    assert (saturate 128 = 127)
    printfn "  saturate: OK"

    let fc = [{ Lo = 0; Hi = 100; Name = "test" }]
    assert (check fc 50).Passed
    assert not (check fc 150).Passed
    printfn "  check: OK"

    let fc4 = [{Lo=0;Hi=10;Name="a"};{Lo=0;Hi=10;Name="b"};{Lo=0;Hi=10;Name="c"};{Lo=0;Hi=10;Name="d"}]
    let r = check fc4 50
    assert (r.Severity = Severity.Critical && r.ViolatedCount = 4)
    printfn "  severity: OK"

    printfn "  All tests pass"
    0
